### PR TITLE
s390x: add FBA-DASD disks support

### PR DESCRIPTION
**Pull request by nikita-dubrovskii**
There are 2 types of DASD devices:
- ECKD (Extended Count Key Data) - is regular DASD of type 3390
- FBA (Fixed Block Access) - is used for emulated device that represents a real SCSI device

Only ECKD-DASD disks require `dasdfmt, fdasd` linux tools to be configured.

Signed-off-by: Nikita Dubrovskii <nikita@linux.ibm.com>

### CODE REVIEW

File: src/install.rs [hunk 1]
```diff
@@ -502,7 +502,18 @@ fn stash_saved_partitions(disk: &mut File, saved: &SavedPartitions) -> Result<()
 fn is_dasd(config: &InstallConfig) -> Result<bool> {
     let target = canonicalize(&config.device)
         .chain_err(|| format!("getting absolute path to {}", config.device))?;
-    Ok(target.to_string_lossy().starts_with("/dev/dasd"))
+    if !target.to_string_lossy().starts_with("/dev/dasd") {
```
Comment (lucab) [inline src/install.rs:505]: Minor: you can move this `.to_string_lossy()` directly into `target` definition, that should spare you the `.display()` a few line below (same applies to `name` too).

Comment (nikita-dubrovskii) [reply src/install.rs:505]: Did it for `name`. From `target` have to extract `file_name`, so couldn't convert it to string

Comment (lucab) [reply src/install.rs:505]: Oh indeed, sorry I missed that aspect.

File: src/install.rs [hunk 2]
```diff
@@ -502,7 +502,19 @@ fn stash_saved_partitions(disk: &mut File, saved: &SavedPartitions) -> Result<()
 fn is_dasd(config: &InstallConfig) -> Result<bool> {
```
Comment (jlebon) [inline src/install.rs:502]: WDYT about renaming this function to `is_eckd_dasd`?

File: src/install.rs [hunk 3]
```diff
@@ -502,7 +502,19 @@ fn stash_saved_partitions(disk: &mut File, saved: &SavedPartitions) -> Result<()
...
+        return Ok(false);
+    }
+    let name = target
+        .file_name()
+        .chain_err(|| format!("getting name of {}", target.display()))?
+        .to_string_lossy();
+    let devtype = format!("/sys/class/block/{}/device/devtype", name);
+    let devtype = std::fs::read_to_string(&devtype).chain_err(|| format!("reading {}", devtype))?;
+    if !devtype.starts_with("3390/") {
+        bail!("only ECKD-DASD (type 3390) devices are supported");
```
Comment (jlebon) [inline src/install.rs:515]: Shouldn't we return `Ok(false)` here? Or does coreos-installer currently not work at all with FBA?

File: src/install.rs (hunk 2 above)
Comment (nikita-dubrovskii) [reply src/install.rs:502]: i don't have a strong opinion on that. it's a first patch in FBA-story, so once i figure out how to configure&use FBA, i would modify this function again

File: src/install.rs (hunk 3 above)
Comment (nikita-dubrovskii) [reply src/install.rs:515]: Currently FBA is not supported - it cannot (and shoudn't ) be configured as ECKD. Regular installation fails as well, so for now it's better to beak installation at this point.

**Review by jlebon — DISMISSED** [review]

**Comment by miabbott** [comment]
@nikita-dubrovskii Does this address any of the s390x BZs that we've seen recently?

**Comment by nikita-dubrovskii** [comment]
@miabbott 
> Does this address any of the s390x BZs that we've seen recently?

there is one internal GitHub-issue  within IBM, but i'm not aware of any BZ. this is a new feature

File: src/s390x/fba.rs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        Ok(())
+    }
+
+    fn make_partitions(&self, device: &mut File, first_mb: &[u8]) -> Result<Vec<Range>> {
+        let bytes_per_block = get_sector_size(&device).unwrap().get().into();
+        let partitions = partitions_from_header(bytes_per_block, first_mb)?;
+        let id: u32 = rand::thread_rng().gen();
+        let mut ranges = Vec::new();
+        let mut mbr = MBR::new_from(device, bytes_per_block as u32, id.to_be_bytes())
+            .chain_err(|| format!("creating new partition table on {}", self.dasd))?;
```
Comment (bgilbert) [inline src/s390x/fba.rs:45]: `creating new partition table for {}`.  This call doesn't write to the disk.

File: src/s390x/fba.rs [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        for (idx, it) in partitions.iter().enumerate() {
+            let size = it.ending_lba - it.starting_lba + 1;
```
Comment (bgilbert) [inline src/s390x/fba.rs:48]: s/size/blocks/ ?

File: src/s390x/fba.rs [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+    fn make_partitions(&self, device: &mut File, first_mb: &[u8]) -> Result<Vec<Range>> {
+        let bytes_per_block = get_sector_size(&device).unwrap().get().into();
+        let partitions = partitions_from_header(bytes_per_block, first_mb)?;
+        let id: u32 = rand::thread_rng().gen();
+        let mut ranges = Vec::new();
+        let mut mbr = MBR::new_from(device, bytes_per_block as u32, id.to_be_bytes())
+            .chain_err(|| format!("creating new partition table on {}", self.dasd))?;
+
+        for (idx, it) in partitions.iter().enumerate() {
```
Comment (bgilbert) [inline src/s390x/fba.rs:47]: `it`?  The ECKD code uses `pt`.

File: src/s390x/fba.rs [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+                out_offset: offset,
+                length: size * bytes_per_block,
+            });
+            mbr[idx + 1] = MBRPartitionEntry {
+                boot: false,
+                first_chs: CHS::empty(),
+                sys: 0x83, // MBR_LINUX_DATA_PARTITION
+                last_chs: CHS::empty(),
+                starting_lba: it.starting_lba as u32,
+                sectors: size as u32,
```
Comment (bgilbert) [inline src/s390x/fba.rs:61]: Let's use `try_into()` on these two lines to catch any overflow from a malformed image.

**Review by bgilbert — COMMENTED** [review]
Some small fixes and one significant refactor, but the functionality looks good.  Thanks for your patience.

File: src/s390x/dasd.rs [hunk 8]
```diff
@@ -120,7 +121,6 @@ pub fn image_copy_s390x(
             .chain_err(|| "copying partition")?;
         cursor += range.length;
     }
-
```
Comment (bgilbert) [inline src/s390x/dasd.rs:107]: Stray deleted line.

**Review by bgilbert — APPROVED** [review]

**Review by bgilbert — APPROVED** [review]

File: src/s390x/eckd.rs [hunk 9]
```diff
@@ -15,194 +15,196 @@
...
+    /// Generate partition table entries and byte ranges to copy
+    fn partition_ranges(
+        &self,
+        device: &mut File,
+        first_mb: &[u8],
+    ) -> Result<(Vec<Range>, Vec<String>)> {
+        let bytes_per_block: u64 = get_sector_size(device).unwrap().get().into();
+        let blocks_per_track: u64 = get_sectors_per_track(device).unwrap().get().into();
+        let partitions = partitions_from_header(bytes_per_block, first_mb)?;
+        let last = partitions.len() - 1;
```
[the remark this answers is not in the document]
Comment (nikita-dubrovskii) [reply src/s390x/eckd.rs]: imho using `last` makes code more concise

**Pull request merged**